lib/array.js

```diff
@@ -171,15 +174,20 @@ const EMPTY_ARR = 'Metasync: reduce of empty array with no initial value';
 const reduce = (items, fn, done, initial) => {
   done = done || common.emptyness;
   const len = items.length;
-  let count = typeof initial === 'undefined' ? 1 : 0;
+  const hasInitial = typeof initial !== 'undefined';
 
-  if (!len) {
-    const err = count ? new TypeError(EMPTY_ARR) : null;
-    done(err, initial);
+  if (len === 0 && !hasInitial) {
+    done(new TypeError(REDUCE_EMPTY_ARR), initial);
     return;
   }
 
-  let previous = count === 1 ? items[0] : initial;
+  let previous = hasInitial ? initial : items[0];
+  if ((len === 0 && hasInitial) || (len === 1 && !hasInitial)) {
+    done(null, previous);
+    return;
+  }
+
+  let count = 1 - hasInitial;
```

Comment:
I am strongly against implicit boolean to number conversion.

Reply:
Done